Add ability to parce brace enclosed tokens

StoyanDimitrov opened the pull request:
Currently is not possible to find `[NS("Strings")]`. This change allows just that.

Here is [an example](/minetest-mods/character_creator/blob/084b67d212717df9573dbe816630af423b02d4ac/skins.lua#L5) translation source.

StoyanDimitrov committed:
Add ability to parce brace enclosed tokens

Currently is not possible to find `[NS("Strings")]`. This change allows just that.

appgurueu commented:
* Indexing a table with translated strings seems like a bit of a potential code smell. Why are you doing it?
* The comment says "look for beginning of file or anything that isn't a function identifier". The purpose of this is to prevent something like `FOOBARS("hello")` from being matched. I think the proper way to fix the regex is to implement just that via a negative lookbehind, so `(?<![a-zA-Z0-9])`, rather than to add many special cases. Is there a problem with this approach that I'm missing?
* Irrelevant for this PR, but for future reference: We probably should look into employing a proper Lua tokenizer and/or parser.

StoyanDimitrov commented:
I'm not a maintainer of the said mod and wanted to contribute just the translation with minimal code changes. That said probably gettext would be better approach (I have a branch locally translated with gettext).
For the change I was considering negative lookbehind but still didn't wanted to change the things too much not having the broader look at the project. If you thing that will be of greater value to the project I'll do it.

I totally understand that the reason for the change is actually not optimal (for translation) structure of the source language.

appgurueu commented:
> For the change I was considering negative lookbehind but still didn't wanted to change the things too much not having the broader look at the project. If you think that will be of greater value to the project I'll do it.

I think that's better, yes. The problem is that making many locally seemingly "sensible" changes is how we end up with a confusing regex which doesn't quite match the explanation. So at some point we should do a little bit of "refactoring" instead of making the situation worse by choosing the path of least resistance. It's still a single-line and conceptually quite simple code change.

StoyanDimitrov committed:
Implement negative lookbehind instead of simple character exclusion

Tested against
```
NS("string1 must present")
must.present.NS("string2 must present")
[NS("string3 must present")]
(NS("string4 must present"))
{NS("string5 must present")}
FAIL_NS("string6 must not present")
false=NS("string7 must present")
,NS("string8 must present")
```

StoyanDimitrov commented:
I allowed myself to include `_` as possibly useful character along with alphanums you requested.

appgurueu commented:
> I allowed myself to include `_` as possibly useful character along with alphanums you requested.

Ah yes, good catch.

appgurueu commented:
Sorry for forgetting about this (I wanted to give others a chance to look at it, I should've merged it directly instead).

I've now merged it as https://github.com/luanti-org/modtools/commit/51beb5ff23c939a04da559470371bc6dbb800bca.
